flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java

```diff
@@ -306,51 +306,88 @@ public void testProducerFailedException() throws Exception {
 	}
 
 	/**
-	 * Tests to verify that the input channel requests floating buffers from buffer pool
-	 * in order to maintain backlog + initialCredit buffers available once receiving the
-	 * sender's backlog, and registers as listener if no floating buffers available.
+	 * Tests to verify that the input channel requests floating buffers from buffer pool for
+	 * maintaining (backlog + initialCredit) available buffers once receiving the sender's backlog.
+	 *
+	 * <p>Verifies the logic of recycling floating buffer back into the input channel and the logic
+	 * of returning extra floating buffer into the buffer pool during recycling exclusive buffer.
 	 */
 	@Test
-	public void testRequestFloatingBufferOnSenderBacklog() throws Exception {
+	public void testRequestAndReturnFloatingBuffer() throws Exception {
 		// Setup
-		final NetworkBufferPool networkBufferPool = new NetworkBufferPool(12, 32, MemoryType.HEAP);
+		final NetworkBufferPool networkBufferPool = new NetworkBufferPool(14, 32, MemoryType.HEAP);
+		final int numExclusiveBuffers = 2;
+		final int numFloatingBuffers = 12;
+
 		final SingleInputGate inputGate = createSingleInputGate();
 		final RemoteInputChannel inputChannel = createRemoteInputChannel(inputGate);
+		inputGate.setInputChannel(inputChannel.partitionId.getPartitionId(), inputChannel);
 		try {
-			final int numFloatingBuffers = 10;
 			final BufferPool bufferPool = spy(networkBufferPool.createBufferPool(numFloatingBuffers, numFloatingBuffers));
 			inputGate.setBufferPool(bufferPool);
-
-			// Assign exclusive segments to the channel
-			final int numExclusiveBuffers = 2;
-			inputGate.setInputChannel(inputChannel.partitionId.getPartitionId(), inputChannel);
 			inputGate.assignExclusiveSegments(networkBufferPool, numExclusiveBuffers);
 
-			assertEquals("There should be " + numExclusiveBuffers + " buffers available in the channel",
-				numExclusiveBuffers, inputChannel.getNumberOfAvailableBuffers());
+			// Prepare the exclusive and floating buffers to verify recycle logic later
+			Buffer exclusiveBuffer = inputChannel.requestBuffer();
+			assertNotNull(exclusiveBuffer);
+			Buffer floatingBuffer1 = bufferPool.requestBuffer();
+			assertNotNull(floatingBuffer1);
+			Buffer floatingBuffer2 = bufferPool.requestBuffer();
+			assertNotNull(floatingBuffer2);
 
-			// Receive the producer's backlog
+			// Receive the producer's backlog less than the number of available floating buffers
 			inputChannel.onSenderBacklog(8);
 
-			// Request the number of floating buffers by the formula of backlog + initialCredit - availableBuffers
-			verify(bufferPool, times(8)).requestBuffer();
+			// Request the floating buffers to maintain (backlog + initialCredit) available buffers
```

Comment:
add note that one exclusive buffer is taken (and therefore 11 requests and not 10)